File: python_lang/src/mathematics/matrix.py
import math


def cholesky_decomposition(mat: list[list]):
    n = len(mat)
    l = [[0] * n for _ in range(n)]
    for i in range(0, n):
        square = mat[i][i]
        for j in range(i):
            square -= l[i][j] ** 2
        l[i][i] = math.sqrt(square)
        for j in range(i + 1, n):
            product = mat[i][j]
            for k in range(j - 1):
                product -= l[i][k] * l[j][k]
            l[j][i] = 0 if l[i][i] == 0 else product / l[i][i]
    for row in l:
        print(row)
    return l


if __name__ == '__main__':
    matrix = [
        [3, -1, 0, -1, 0],
        [-1, 3, -1, 3, -1],
        [0, -1, 3, 1, 0],
        [-1, 3, 1, 5, -1],
        [0, -1, 0, -1, 1]
    ]
    cholesky_decomposition(matrix)
# [1.7320508075688772, 0, 0, 0, 0]
# [-0.5773502691896258, 1.632993161855452, 0, 0, 0]
# [0.0, -0.6123724356957945, 1.620185174601965, 0, 0]
# [-0.5773502691896258, 1.632993161855452, 1.2344267996967353, 0.6900655593423545, 0]
# [0.0, -0.6123724356957945, -0.23145502494313783, 0.4140393356054122, 0.6324555320336761]

File: r_lang/src/base/matrix.R
library(Matrix)
require(Matrix)

A1 <- matrix(c(2, -1, 0, -1, 0,
               -1, 3, -1, 3, -1,
               0, -1, 3, 1, 0,
               -1, 3, 1, 6, -1,
               0, -1, 0, -1, 1), nrow = 5, byrow = TRUE)
t(chol(A1))
A2 <- matrix(c(2, -1, 0, -1,
               -1, 3, -1, 3,
               0, -1, 3, 1,
               -1, 3, 1, 6), nrow = 4, byrow = TRUE)
t(chol(A2))

cholesky_decomposition <- function(A) {
  if (!isSymmetric.matrix(A)) {
    print('error: parameter A should be a n*n symmetric matrix')
    return()
  }
  n <- nrow(A)
  for (i in 1:n) {
    if (A[i, i] < 0) {
      print(sprintf('error: the leading minor of order %d is not positive', i))
      return()
    }
  }
  L <- Matrix(0, n, n)
  for (i in 1:n) {
    square <- A[i, i]
    if (1 <= i) {
      for (j in 1:i) {
        square <- square - (L[i, j]^2)
      }
    }
    if (square < 0) {
      print(sprintf('error: the leading minor of order %d is not positive', i))
      return()
    }
    L[i, i] <- sqrt(square)
    if ((i + 1) <= n) {
      for (j in (i + 1):n) {
        product <- A[i, j]
        if (1 <= (j - 1)) {
          for (k in 1:(j - 1)) {
            product <- product - L[i, k] * L[j, k]
          }
        }
        if ((L[i, i] == 0) & product != 0) {
          print(sprintf('error: the leading minor of order %d is not positive', i))
          return()
        }
        L[j, i] <- product / L[i, i]
      }
    }
  }
  L
}

L <- cholesky_decomposition(A2); L
L %*% t(L)

CholeskyDecompositionTest <- function() {
  error.count <- 0
  for (mtx in list(
    matrix(c(2, -1, 0, -1,
             -1, 3, -1, 3,
             0, -1, 3, 1,
             -1, 3, 1, 6), nrow = 4, byrow = TRUE),
    matrix(c(2, -1, 0, -1, 0,
             -1, 3, -1, 3, -1,
             0, -1, 3, 1, 0,
             -1, 3, 1, 6, -1,
             0, -1, 0, -1, 1), nrow = 5, byrow = TRUE)
  )) {
    L <- cholesky_decomposition(mtx)
    mtx.new <- L %*% t(L)
    bias <- 0.1^6
    n <- ncol(mtx)
    if (sum(abs(mtx.new - mtx) > bias)) {
      print('different results:')
      print(mtx)
      print(mtx.new)
      error.count <- error.count + 1
    }
  }
  if (error.count) {
    print(sprintf('CholeskyDecomposition function test failed: %d testcase.', error.count))
  }else {
    print('CholeskyDecomposition function test passed.')
  }
}

CholeskyDecompositionTest()